### PR TITLE
Optimize json generation

### PR DESCRIPTION
**Pull request by casualjim**
We are using this in an on-box agent that is very busy, we were asked to reduce the CPU usage.

from @hariharan-a :

> With your spectator-go fixes. Definitely a lot better. %usr cpu is between 0.7% - 0.10%. It used to be 0.23% - 0.40% without the fix.

### CODE REVIEW

File: http_client.go [hunk 1]
```diff
@@ -34,35 +37,66 @@ func userFriendlyErr(errStr string) string {
...
+	payloadBuffer := payloadPool.Get().(*bytes.Buffer)
+	var g *gzip.Writer
 	if compressed {
-		payloadBuffer = &bytes.Buffer{}
-		g := gzip.NewWriter(payloadBuffer)
+		g = gzipWriterPool.Get().(*gzip.Writer)
+		g.Reset(payloadBuffer)
+		defer func() {
+			if err := g.Close(); err != nil {
+				log.Printf("closing gzip writer: %v", err)
```
Comment (brharrington) [inline http_client.go:64]: Logging should be done with the logger configured on the registry. These should probably be at debug level.

**Review by brharrington — CHANGES_REQUESTED** [review]
Thanks for the contribution.

For context, we typically try to limit the dependencies for Spectator libraries to the standard library for a language to minimize conflicts and compatibility issues for users. This has been a big pain point in the past. @dmuino is more familiar with Go and should be able to take a look next week.

**Comment by casualjim** [comment]
I updated the logger usage

**Review by dmuino — APPROVED** [review]
The changes look good to me. As Brian mentioned a goal with these libraries is to reduce the number of dependencies to a minimum so they're easier to integrate into codebases that might have conflicting requirements. With go modules there should be straightforward workarounds in that case so I think we can just merge this PR.

Note that our plan is to migrate our thin clients to just relay their activity to [spectatord](https://github.com/Netflix-Skunkworks/spectatord) and then focus all our optimizations in that project.

File: http_client.go [hunk 2]
```diff
@@ -34,35 +36,66 @@ func userFriendlyErr(errStr string) string {
...
+	payloadBuffer := payloadPool.Get().(*bytes.Buffer)
+	var g *gzip.Writer
 	if compressed {
-		payloadBuffer = &bytes.Buffer{}
-		g := gzip.NewWriter(payloadBuffer)
+		g = gzipWriterPool.Get().(*gzip.Writer)
+		g.Reset(payloadBuffer)
+		defer func() {
+			if err := g.Close(); err != nil {
+				log.Infof("closing gzip writer, best effort: %v", err)
```
Comment (brharrington) [inline http_client.go:64]: I wonder if this should be debug level, not sure it is useful to see it every 5s in most cases.

**Review by brharrington — APPROVED** [review]

File: http_client.go (hunk 2 above)
Comment (casualjim) [reply http_client.go]: I agree, I updated it

**Pull request merged**